BookingsApi/BookingsApi.DAL/Services/HearingAllocationService.cs

```diff
@@ -0,0 +1,204 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using BookingsApi.Common.Configuration;
+using BookingsApi.Common.Helpers;
+using BookingsApi.Common.Services;
+using BookingsApi.Domain;
+using BookingsApi.Domain.Enumerations;
+using BookingsApi.Domain.Validations;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Options;
+
+namespace BookingsApi.DAL.Services
+{
+    public interface IHearingAllocationService
+    {
+        Task<JusticeUser> AllocateAutomatically(Guid hearingId);
+    }
+
+    public class HearingAllocationService : IHearingAllocationService
+    {
+        private readonly BookingsDbContext _context;
+        private readonly IRandomNumberGenerator _randomNumberGenerator;
+        private readonly AllocateHearingConfiguration _configuration;
+
+        public HearingAllocationService(
+            BookingsDbContext context, 
+            IRandomNumberGenerator randomNumberGenerator,
+            IOptions<AllocateHearingConfiguration> configuration)
+        {
+            _context = context;
+            _randomNumberGenerator = randomNumberGenerator;
+            _configuration = configuration.Value;
+        }
+        
+        public async Task<JusticeUser> AllocateAutomatically(Guid hearingId)
+        {
+            var hearing = await _context.VideoHearings.SingleOrDefaultAsync(x => x.Id == hearingId);
+            if (hearing == null)
+            {
+                throw new DomainRuleException("HearingNotFound",
+                    $"Hearing {hearingId} not found");
+            }
+
+            var allocations = _context.Allocations.Where(a => a.HearingId == hearing.Id).ToList();
+            if (allocations.Any())
+            {
+                throw new DomainRuleException("HearingAlreadyAllocated",
+                    $"Hearing {hearing.Id} has already been allocated");
+            }
+
+            var hearingStartTime = hearing.ScheduledDateTime;
+            var hearingEndTime = hearing.ScheduledEndTime;
+
+            if (hearingStartTime.Date != hearingEndTime.Date)
+            {
+                throw new DomainRuleException("AllocationNotSupported",
+                    $"Unable to allocate to hearing {hearing.Id}, hearings which span multiple days are not currently supported");
+            }
+
+            var cso = SelectCso(hearingStartTime, hearingEndTime);
+            if (cso == null)
+            {
+                throw new DomainRuleException("NoCsosAvailable",
+                    $"Unable to allocate to hearing {hearingId}, no CSOs available");
+            }
+
+            return cso;
+        }
+
+        private JusticeUser SelectCso(DateTime hearingStartTime, DateTime hearingEndTime)
+        {
+            var availableCsos = GetAvailableCsos(hearingStartTime, hearingEndTime);
```

Comment:
I think because GetAvailableCsos returns IEnumerable, availableCsos is lazy loaded; so every time it's used below it has to iterate over all the values. Best practice is to return an array I think, or do GetAvailableCsos.ToArray()

Reply:
Because GetAvailableCsos already loads from the db using ToList(), I don't think there should be any lazy loading. I can change it to return a List<JusticeUser>, maybe that would be clearer.